src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_fsm.h

```diff
@@ -0,0 +1,178 @@
+    {
+        using namespace boost::sml;
+
+        const auto block_s          = state<BlockState>;
+        const auto steal_and_chip_s = state<StealAndChipState>;
+
+        const auto update_e = event<Update>;
+
+        /**
+         * Guard that checks if the enemy threat has ball
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_has_ball = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return enemy_threat_opt.value().has_ball;
+            };
+            return false;
+        };
+
+        /**
+         * Guard that checks if the enemy threat is near
+         * our robot
+         *
+         * @param event ShadowEnemyFSM::Update
+         *
+         * @return if the ball has been have_possession
+         */
+        const auto enemy_threat_is_near = [](auto event) {
+            std::optional<EnemyThreat> enemy_threat_opt =
+                event.control_params.enemy_threat;
+            if (enemy_threat_opt.has_value())
+            {
+                return distance(enemy_threat_opt.value().robot.position(),
+                                event.common.robot.position()) <=
+                       event.control_params.steal_and_chip_distance;
+            };
+            return false;
+        };
+
+        /**
+         * Action to block the ball
+         *
+         * If the enemy can pass to the robot we are shadowing, block the pass.
+         * Else, block the net
+         *
+         * @param event ShadowEnemyFSM::Update
+         */
+        const auto block_pass_or_net = [](auto event) {
+            auto ball_position = event.common.world.ball().position();
+            auto face_ball_orientation =
+                (ball_position - event.common.robot.position()).orientation();
+            Point position_to_block;
+            if (!event.control_params.enemy_threat.value().has_ball)
+            {
+                Vector enemy_to_passer_vector =
+                    ball_position -
+                    event.control_params.enemy_threat.value().robot.position();
+
+                position_to_block =
+                    event.control_params.enemy_threat.value().robot.position() +
+                    enemy_to_passer_vector.normalize(
+                        event.control_params.shadow_distance);
+            }
+            else
+            {
+                std::vector<Robot> robots_to_ignore = {event.common.robot};
+                if (event.common.world.friendlyTeam().goalie().has_value())
+                {
+                    robots_to_ignore.emplace_back(
+                        event.common.world.friendlyTeam().goalie().value());
+                }
+
+                auto best_enemy_shot_opt = calcBestShotOnGoal(
```

Comment:
There's a larger question here on whether the shadower or the goalie should look at calcbestshotongoal that we should have outside of this PR